gfx.cpp

```diff
@@ -1066,45 +1070,60 @@ class GfxInternal
-            uint32_t adapter_scores[ARRAYSIZE(adapters)] = {};
-            DXGI_ADAPTER_DESC1 adapter_descs[ARRAYSIZE(adapters)] = {};
-            for(uint32_t i = 0; i < ARRAYSIZE(adapters); ++i)
+            if(SUCCEEDED(factory->QueryInterface(IID_PPV_ARGS(&factory6))))
             {
-                IDXGIAdapter1 *adapter1 = nullptr;
-                if(!SUCCEEDED(factory->EnumAdapters1(i, &adapter1)))
-                    break;
-                uint32_t j, adapter_score;
-                adapter1->GetDesc1(&adapter_desc);
-                if((adapter_desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE) != 0)
-                    adapter_score = 0;
-                else
-                    switch(adapter_desc.VendorId)
-                    {
-                    case 0x1002u:   // AMD
-                        adapter_score = 3;
-                        break;
-                    case 0x10DEu:   // NVIDIA
-                        adapter_score = 2;
+                for(uint32_t i = 0; i < ARRAYSIZE(adapters); ++i)
+                {
+                    IDXGIAdapter1 *adapter1 = nullptr;
+                    if(!SUCCEEDED(factory6->EnumAdapterByGpuPreference(i, DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE, IID_PPV_ARGS(&adapter1))))
                         break;
-                    default:
-                        adapter_score = 1;
+                    adapters[i] = adapter1;
+                }
+            }
+            else
+            {
+                DXGI_ADAPTER_DESC1 adapter_desc = {};
+                uint32_t adapter_scores[ARRAYSIZE(adapters)] = {};
+                DXGI_ADAPTER_DESC1 adapter_descs[ARRAYSIZE(adapters)] = {};
+                for(uint32_t i = 0; i < ARRAYSIZE(adapters); ++i)
+                {
+                    IDXGIAdapter1 *adapter1 = nullptr;
+                    if(!SUCCEEDED(factory->EnumAdapters1(i, &adapter1)))
                         break;
+                    uint32_t j, adapter_score;
+                    adapter1->GetDesc1(&adapter_desc);
+                    if((adapter_desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE) != 0)
+                        adapter_score = 0;
+                    else
+                        switch(adapter_desc.VendorId)
+                        {
+                        case 0x1002u:   // AMD
+                            adapter_score = 2;
+                            break;
+                        case 0x10DEu:   // NVIDIA
+                            adapter_score = 2;
+                            break;
+                        default:
+                            adapter_score = 1;
+                            break;
+                        }
+                    for(j = 0; j < i; ++j)
+                        if(adapter_score > adapter_scores[j] ||
+                           adapter_desc.DedicatedVideoMemory > adapter_descs[j].DedicatedVideoMemory ||
+                           adapter_desc.SharedSystemMemory > adapter_descs[j].SharedSystemMemory)
```

Comment:
Also extended the memory check to also order by shared memory which is a fallback for laptop GPUs